### PR TITLE
Fix `F.softmax_cross_entropy` float16 under/overflow

### PR DESCRIPTION
**Pull request by niboshi**
Depends on #6363.

### CODE REVIEW

**Comment by niboshi** [comment]
Rebased.

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 1]
```diff
@@ -140,6 +151,14 @@ def forward(self, inputs, device):
         if not (self.enable_double_backprop or device.xp is chainerx):
             assert (loss.creator.y is not None) == self.cache_score
 
+        # All the loss values except those corresponding to the ignored label
+        # must be positive.
+        # TODO(niboshi): Use device.xp.where once chainerx supports it.
+        assert numpy.where(
+            _cpu._to_cpu(t == -1),
+            True,
+            _cpu._to_cpu(loss.array) > 0).all()
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Why not use the public API?

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 2]
```diff
@@ -328,37 +362,41 @@ def _double_backward_softmax_cross_entropy(x, t, normalize, class_weight,
...
             count = len(x)
         count = max(count, 1.)
-        loss = loss / count
-        return chainer.functions.sum(loss)
-    else:
-        return loss
+        # Perform reduction in a promoted dtype
+        loss = F.cast(loss, reduc_dtype)
+        loss = F.sum(loss / count)
+        loss = F.cast(loss, x.dtype)
```
Comment (hvy) [inline chainer/functions/loss/softmax_cross_entropy.py]: How about skipping these calls to `F.cast` in case dtypes are the same?

**Comment by hvy** [comment]
Sorry for the late review, I wrote some comments. Could you take a look?

File: chainer/functions/loss/softmax_cross_entropy.py (hunk 2 above)
Comment (niboshi) [reply chainer/functions/loss/softmax_cross_entropy.py]: Fixed.

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py (hunk 1 above)
Comment (niboshi) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Fixed.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Travis failure seems unrelated.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
LGTM!
